subsys/net/ip/tcp.c

```diff
@@ -1201,8 +1201,6 @@ static void tcp_resend_data(struct k_work *work)
 	conn->data_mode = TCP_DATA_MODE_RESEND;
 	conn->unacked_len = 0;
 
-	(void)k_sem_take(&conn->tx_sem, K_NO_WAIT);
```

Comment:
Don't we need to remove the related `k_sem_give()` too as the semaphore calls usually come in pairs?

Reply:
The conn->tx_sem is used to indicate if transmission is allowed. Generally this is blocked when the transmit window is full. There was a special case when the connection entered re-transmit mode, where also the semaphore was blocked.
The semaphore is freed upon reception of a TCP-ACK for the re-transmitted data, which should still persist.

Reply:
I have always found it a strange line as well, but as I understand it the semaphore is taken at either a window full of retransmission. As soon as an ACK comes in the retransmission will end or data is release from the send_data buffer and the semaphore is given.

@rlubos: Can you comment on this one, as you inserted this line in https://github.com/zephyrproject-rtos/zephyr/commit/7eaacdaca60f4242171ed4bd4d334bac2cb7675d?

Reply:
Actually, that is a good point, but I think we should rather go the other way around, and remove that `k_sem_give()` below, since it seems no longer needed. 

Previously it was assumed that once TCP stack entered retransmit mode, the semaphore was cleared, as the application was not able to queue new data when TCP was in retransmit mode. After the changes by @ssharks though, it is no longer the case, so it's no longer needed to clear the sem once we enter this mode, but at the same time it's no longer needed to give it back once retransmit mode is left.

Reply:
Thanks for your critical review, I totally missed the statement in line 1231. Although it probably does not hurt, it is cleaner to remove this statement.